src/Couchbase.Lite.Shared/API/Sync/ReplicatorConfiguration.cs

```diff
@@ -73,6 +76,230 @@ public enum DocumentFlags
+        /// </exception>
+        public TimeSpan? MaxAttemptsWaitTime
+        {
+            get => Options.MaxAttemptsWaitTime;
+            init => Options.MaxAttemptsWaitTime = value;
+        }
+
+        /// <summary>
+        /// Gets the target to replicate with (either <see cref="Database"/>
+        /// or <see cref="Uri"/>
+        /// </summary>
+        [NotNull]
+        public IEndpoint Target { get; }
+
+        /// <summary>
+        /// The implemented custom conflict resolver object can be registered to the replicator 
+        /// at ConflictResolver property. The default value of the conflictResolver is null. 
+        /// When the value is null, the default conflict resolution will be applied.
+        /// </summary>
+        [CanBeNull]
+        public IConflictResolver ConflictResolver { get; init; }
+
+        internal TimeSpan CheckpointInterval
+        {
+            get => Options.CheckpointInterval;
+            init => Options.CheckpointInterval = value;
+        }
+
+        [NotNull]
+        internal ReplicatorOptionsDictionary Options { get; set; } = new ReplicatorOptionsDictionary();
+
+        [CanBeNull]
+        internal Database OtherDB { get; init; }
+
+        [CanBeNull]
+        internal Uri RemoteUrl { get; init; }
+
+        internal C4SocketFactory SocketFactory
+        {
+            get => _socketFactory.open != IntPtr.Zero ? _socketFactory : LiteCore.Interop.SocketFactory.InternalFactory;
+            init => _socketFactory = value;
+        }
+
+        #endregion
+
+        #region Constructors
+
+        /// <summary>
+        /// Constructs a new builder object with the required properties
+        /// </summary>
+        /// <param name="database">The database that will serve as the local side of the replication</param>
+        /// <param name="target">The endpoint to replicate to, either local or remote</param>
+        /// <exception cref="ArgumentException">Thrown if an unsupported <see cref="IEndpoint"/> implementation
+        /// is provided as <paramref name="target"/></exception>
+        public ReplicatorConfiguration([NotNull] Database database, [NotNull] IEndpoint target)
+        {
+            Database = CBDebug.MustNotBeNull(WriteLog.To.Sync, Tag, nameof(database), database);
+            Target = CBDebug.MustNotBeNull(WriteLog.To.Sync, Tag, nameof(target), target);
+
+            if(target.GetType() == typeof(URLEndpoint)) {
+                RemoteUrl = ((URLEndpoint)target).Url;
+            }
+            #if COUCHBASE_ENTERPRISE
+            else if(target.GetType() == typeof(DatabaseEndpoint)) {
+                OtherDB = ((DatabaseEndpoint)target).Database;
+            } else if (target.GetType() == typeof(MessageEndpoint)) {
+                _socketFactory = ((MessageEndpoint)target).SocketFactory;
+                RemoteUrl = new Uri("x-msg-endpoint://");
+                Options.RemoteDBUniqueID = ((MessageEndpoint)target).Uid;
+            }
```

Comment:
If this is what it is going to take to make things work with internal init-only properties than I changed my mind and the old way is better.  This is leaking too much EE implementation stuff into here.

Reply:
Agree, and the old way looks much cleaner. Reverted.